### PR TITLE
Make Quarkus Kubernetes GitOps friendly

### PR DESCRIPTION
**Pull request by Dairdevil**
Done by not writing to target dir and removing non-idempotent annotations.

Addressing issue: https://github.com/quarkusio/quarkus/issues/26928

### CODE REVIEW

**Review by iocanel — COMMENTED** [review]
The pull request looks good to me.

I haven't approved though as I am currently on PTO and I think that we mught need to decide on the timing of this change first (as this changes default behavior).

I am wondering if internally we should use the term `idempotent` vs gitops firendly as the `git ops friendly` may not be obvious to some.

**Comment by Dairdevil** [comment]
Thanks @iocanel - good suggestion on using the term "idempotent", I'll change it to that. It was a case of "I know this isn't a good label for it but I'm having a mind blank as to what would be better".

Would it be preferable to default to the current behaviour & leave the change of default behaviour to be done later?

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessorUtil.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+public class KubernetesProcessorUtil {
+
+    public static boolean makeGitOpsFriendly() {
+        Config config = ConfigProvider.getConfig();
+        String configValue = config.getOptionalValue(GITOPS_FRIENDLY, String.class)
+                .orElse("true")
+                .trim();
+
+        if ("n".equalsIgnoreCase(configValue) || "no".equalsIgnoreCase(configValue)) {
```
Comment (gastaldi) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessorUtil.java:19]: `parseBoolean` below already returns `false` for these values

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessorUtil.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.Optional;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+
+public class KubernetesProcessorUtil {
+
+    public static boolean makeGitOpsFriendly() {
+        Config config = ConfigProvider.getConfig();
+        String configValue = config.getOptionalValue(GITOPS_FRIENDLY, String.class)
```
Comment (gastaldi) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessorUtil.java:15]:
```suggestion
        boolean configValue = config.getOptionalValue(GITOPS_FRIENDLY, Boolean.class)
```

This is a better solution, as SmallRye config does the parsing for you (and trimming)

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessorUtil.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                .orElse("true")
```
Comment (gastaldi) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessorUtil.java:16]: If you define a default value of `true` you don't need an optional

Comment (Dairdevil) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessorUtil.java:16]: Where would be the proper place to define this default value?

**Comment by Dairdevil** [comment]
@gastaldi I wanted to avoid making the KubernetesProcessor class take injected configuration as this would break from the other processors. Happy to change this to `@ConfigItem` and remove the static methods if thats preferred?

**Comment by gastaldi** [comment]
Using a Config class with `@ConfigItem` is better because the build generates an Asciidoc document that can be used in the guides

**Comment by Dairdevil** [comment]
Thanks, I'll make that change. Just to make sure I've understood, I should make a new `SomethingConfig` class that has variables of my new config items injected via `@ConfigItem` (as opposed to putting the `@ConfigItem`s directly into the classes they are needed in)?

**Comment by gastaldi** [comment]
@Dairdevil you need to add to https://github.com/quarkusio/quarkus/blob/main/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesConfig.java and use that in your `KubernetesProcessor`

**Comment by Dairdevil** [comment]
@gastaldi I've made the changes based on your explanation, thanks for going through that with me.

I think its as concise as it can be, and currently introduces the following new variables:

- `quarkus.kubernetes.idempotent` Controls whether to include the non-idempotent labels on generated kubernetes resources (defaults to true);
- `quarkus.openshift.idempotent` Controls whether to include the non-idempotent labels on generated openshift resources (defaults to true);
- `quarkus.knative.idempotent` Controls whether to include the non-idempotent labels on generated knative resources (defaults to true);
- `quarkus.kubernetes.output-directory` Optionally write any generated resources to a different directory

Current behaviour is that:

1. By default, all resources are generated with the non-idempotent labels removed
2. Each resource type (kubernetes, openshift, knative) has an independent variable to force the inclusion of the non-idempotent labels
3. If `quarkus.kubernetes.output-directory` is set, new resources are written to this directory in the module root
4. If any of the `quarkus.*.idempotent` variables are set to `false` then the new resources are written to the original location in the `target` directory
5. By default, all new resources will be written to a new directory `.kubernetes` in the module root

This slightly muddled logic has come from tying the writing of the resources to a new directory to the idempotency option. Would it make more sense to completely separate these two and continue to write to the directory under `target` until explicitly specified otherwise (using `quarkus.kubernetes.output-directory` as above) ?

**Comment by gastaldi** [comment]
Looks okay to me, but I'll let @iocanel have the final say on this 👍🏻

**Review by iocanel — APPROVED** [review]
LGTM

**Comment by iocanel** [comment]
@gsmet: The PR does change the default behavior by removing some labels / annotations from the generated resources. Not sure if this qualifies as breaking change, but it might be for a really small portion of our users. I think that we should merge anyway, but just wanted your view on that before I do.

**Comment by gsmet** [comment]
I rebased the PR as somehow the CI didn't run. Let's see how it goes and merge if it's green.

**Comment by Dairdevil** [comment]
@gsmet looks like lots of failing tests because the change has moved where the generated kubernetes are output to by default. Propose that I change the existing ITs application.properties to disable this new feature for all existing tests and then add a new test where it is enabled, what do you think?

**Comment by gsmet** [comment]
@Dairdevil yes it's probably a good idea.

**Comment by Dairdevil** [comment]
@gsmet been a bit busy but looking at this again. Wondering if it will be easier going forward to introduce a new `QuarkusProdModeTest` class that extends the existing but includes Kubernetes specific config and sets the problem value to false? Means not having to include this property in the existing tests or their properties files, just changing the Kubernetes tests to using this new class in place of `QuarkusProdModeTest`. Make sense or am I overthinking it?

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java [hunk 4]
```diff
@@ -182,14 +182,24 @@ public void build(ApplicationInfoBuildItem applicationInfo,
...
                 List<String> generatedFiles = new ArrayList<>(generatedResourcesMap.size());
                 List<String> generatedFileNames = new ArrayList<>(generatedResourcesMap.size());
+
+                Path targetDirectory = outputTarget.getOutputDirectory().resolve(KUBERNETES);
+                Optional<String> overrideOutputDirectory = kubernetesConfig.getOutputDirectory();
+
+                if (overrideOutputDirectory.isPresent()) {
+                    targetDirectory = Paths.get("").toAbsolutePath().resolve(overrideOutputDirectory.get());
+                } else if (KubernetesConfigUtil.isIdempotent(kubernetesConfig, knativeConfig, openshiftConfig)) {
+                    targetDirectory = Paths.get("").toAbsolutePath().resolve(".kubernetes");
```
Comment (Sgitario) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java:192]: I don't see why being idempotent should affect the target directory

**Review by Sgitario — CHANGES_REQUESTED** [review]
I'm missing test coverage for this change.
Also, the annotation `app.quarkus.io/vcs-url` should not be added when idempotent is true. 
Finally, I would rename the property `idempotent` to a more user-friendly name as perhaps not users are familiar with such a concept.

**Comment by Sgitario** [comment]
Ey @Dairdevil, are you still interested in finishing the changes in this pull request?

**Comment by iocanel** [comment]
Rebased the pull request, let's see if tests are passing.

**Comment by iocanel** [comment]
>  Also, the annotation `app.quarkus.io/vcs-url` should not be added when idempotent is true. Finally, I would rename the property `idempotent` to a more user-friendly name as perhaps not users are familiar with such a concept.

I think it's acceptable to assume the the `vcs-url` does not change even among users. Sure, there might be exceptions to the rule, but I think that these are `edge` cases.

`Idempotent` is a term commonly used in Math and Computer Science to describe the concept, so I doubt that there is a more fitting term out there. Do you have a particular term in mind?

**Comment by Sgitario** [comment]
> `Idempotent` is a term commonly used in Math and Computer Science to describe the concept, so I doubt that there is a more fitting term out there. Do you have a particular term in mind?

Yes, that's why I don't like it :)
What about `quarkus.kubernetes.add-vcs-info=true` or `quarkus.kubernetes.add-vcs-data=true`?

**Comment by iocanel** [comment]
> > `Idempotent` is a term commonly used in Math and Computer Science to describe the concept, so I doubt that there is a more fitting term out there. Do you have a particular term in mind?
> 
> Yes, that's why I don't like it :) What about `quarkus.kubernetes.add-vcs-info=true` or `quarkus.kubernetes.add-vcs-data=true`?

To make things gitops friendly, the resources we generate need to be `idempotent`. 
So, the goal is not provide fine grained control over what is added to generated manifest. The goal is that generated manifests are `idempotent`. 

Why? Becasue, if we need to add the generated resource on version control we need to ensure that no matter how many times we run the build the generated resources won't change. Ideally, resources won't change even if we run the build on separte machines. This is best described by the term `idempotent`.

So, what do we need in order to guarantee idempotency?
- avoid adding the commit sha (the vcs-url is not breaking idempotentcy).
- avoid adding timestamps
- avoid using username as a fallback value for `group` and `image` (that's tricky).

Now, if you want to add fine grained control over these properties with individual flags, please be my guest. However, we also need a signle property that can influence all three flags, unless explicitly specified.

tldr; Let's `keep` idempotent. We can optionally add fine grained control.

**Comment by Sgitario** [comment]
Superseeded by https://github.com/quarkusio/quarkus/pull/31373
